src/lib.rs

```diff
@@ -1,66 +1,248 @@
+    assert!(!is_cr('A'));
+    assert!(!is_cr('z'));
+    assert!(!is_cr('Z'));
 }
 
-pub fn is_bit(c: char) -> bool {
-    c == '0' || c == '1'
+/// Controls
+///
+/// CTL = %x00-1F / %x7F
+pub fn is_ctl(c: impl AsChar) -> bool {
+    matches!(c.as_char(), '\x00'..='\x1F' | '\x7F')
 }
 
-pub fn is_char(c: char) -> bool {
-    matches!(c, '\x01'..='\x7F')
+#[test]
+fn test_is_ctl() {
+    assert!(is_ctl(b'\x00'));
+    assert!(is_ctl('\x1F'));
+    assert!(is_ctl('\x7F'));
+    assert!(!is_ctl('a'));
+    assert!(!is_ctl('A'));
+    assert!(!is_ctl('z'));
+    assert!(!is_ctl('Z'));
 }
 
-pub fn is_cr(c: char) -> bool {
-    c == '\r'
+/// 0-9
+///
+/// DIGIT = %x30-39
+pub fn is_digit(c: impl AsChar) -> bool {
+    matches!(c.as_char(), '\x30'..='\x39')
 }
 
-// CRLF
+#[test]
+fn test_is_digit() {
+    assert!(is_digit(b'0'));
+    assert!(is_digit('9'));
+    assert!(!is_digit('a'));
+    assert!(!is_digit('A'));
+    assert!(!is_digit('z'));
+    assert!(!is_digit('Z'));
+}
 
-pub fn is_ctl(c: char) -> bool {
-    matches!(c, '\x00'..='\x1F' | '\x7F')
+/// " (Double Quote)
+///
+/// DQUOTE = %x22
+pub fn is_dquote(c: impl AsChar) -> bool {
+    matches!(c.as_char(), '\x22')
 }
 
-pub fn is_digit(c: char) -> bool {
-    c.is_ascii_digit()
+#[test]
+fn test_is_dquote() {
+    assert!(is_dquote(b'"'));
+    assert!(!is_dquote('\''));
+    assert!(!is_dquote('a'));
+    assert!(!is_dquote('A'));
+    assert!(!is_dquote('z'));
+    assert!(!is_dquote('Z'));
 }
 
-pub fn is_dquote(c: char) -> bool {
-    c == '"'
+/// HEXDIG = DIGIT / "A" / "B" / "C" / "D" / "E" / "F"
+/// ABNF string are case-insensitive so `a` / ... / `f` are allowed
+/// <https://github.com/duesee/abnf-core/issues/12>
+pub fn is_hexdig(c: impl AsChar) -> bool {
+    matches!(c.as_char(), '0'..='9' | 'a'..='f' | 'A'..='F')
 }
 
-pub fn is_hexdig(c: char) -> bool {
-    c.is_ascii_hexdigit()
+#[test]
+fn test_is_hexdige() {
+    assert!(is_hexdig(b'0'));
+    assert!(is_hexdig('9'));
+    assert!(is_hexdig('a'));
+    assert!(is_hexdig('f'));
+    assert!(is_hexdig('A'));
+    assert!(is_hexdig('F'));
+    assert!(!is_hexdig('z'));
+    assert!(!is_hexdig('Z'));
 }
 
-// HTAB
+/// horizontal tab
+///
+/// HTAB = %x09
+pub fn is_htab(c: impl AsChar) -> bool {
+    matches!(c.as_char(), '\x09')
+}
 
-// LF
+#[test]
+fn test_is_htab() {
+    assert!(is_htab(b'\t'));
+    assert!(!is_dquote(' '));
+    assert!(!is_dquote('a'));
+    assert!(!is_dquote('A'));
+    assert!(!is_dquote('z'));
+    assert!(!is_dquote('Z'));
+}
 
-// LWSP
+/// linefeed
+///
+/// LF = %x0A
+pub fn is_lf(c: impl AsChar) -> bool {
+    matches!(c.as_char(), '\x0A')
+}
 
-// OCTET
+#[test]
+fn test_is_lf() {
+    assert!(is_lf(b'\n'));
+    assert!(!is_lf('\r'));
+    assert!(!is_lf('a'));
+    assert!(!is_lf('A'));
+    assert!(!is_lf('z'));
+    assert!(!is_lf('Z'));
+}
 
-// SP
+/// 8 bits of data
+///
+/// OCTET = %x00-FF
+pub fn is_octet(c: impl AsChar) -> bool {
+    matches!(c.as_char(), '\x00'..='\x7F')
```

Comment:
This should be 0x00 to 0xFF (not 0x7F), i.e. any byte. However, 0xFF is not a valid char in Rust and this does not work. What is the alternative?

Reply:
force to take a `u8` here I trough about it, since `octet` rule also have this problem to not make sense for char:

```rust
pub fn us_octet(c: u8) -> bool {
    true
}
```

For completeness.

Reply:
or `matches!(c.as_char(), 0x00 as char ..= 0xFF as char)`